File: haskell/Problem17.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 17
-- Number letter counts

module Problem17 where

letters :: Int -> Int
letters n
    | n >= 0 && n <= 19 = upTo19 n
    | n <= 99 = tens (n `div` 10) + letters (n `mod` 10)
    | n <= 999 && n `mod` 100 == 0 = letters (n `div` 100) + hundred
    | n <= 999 = letters (n `div` 100) + hundred + and + letters (n `mod` 100)
    | n == 1000 = letters 1 + thousand
    | otherwise = 0
  where
    upTo19 = ([0, 3, 3, 5, 4, 4, 3, 5, 5, 4, 3, 6, 6, 8, 8, 7, 7, 9, 8, 8] !!)
    tens = ([0, 3, 6, 6, 5, 5, 5, 7, 6, 6] !!)
    and = 3
    hundred = 7
    thousand = 8

solve :: Int
solve = sum . map letters $ [1..1000]

File: c++/src/problem_17.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 17
// Number letter counts

namespace problem_17 {

long solve() {
	constexpr long and_word = 3;
	constexpr long hundred = 7;
	constexpr long one_thousand = 3 + 8;
	constexpr long a = 0+3+3+5+4+4+3+5+5+4; //  0,  1, ...,  9
	constexpr long b = 3+6+6+8+8+7+7+9+8+8; // 10, 11, ..., 19
	constexpr long c = 6+6+5+5+5+7+6+6;     // 20, 30, ..., 90
	constexpr long s99 = 10*c + 9*a + b;
	constexpr long result =
		10*s99 + (100*a + 9*100*hundred) + 9*99*and_word + one_thousand;
	return result;
}

} // namespace problem_17
